In [1]:
import pandas as pd
import numpy as np
import keras

from keras.models import Sequential
from keras.layers import Dense

concrete_data = pd.read_csv('https://cocl.us/concrete_data')
concrete_data.head(10)

Cement  Blast Furnace Slag  Fly Ash  ...  Fine Aggregate  Age  Strength
0   540.0                 0.0      0.0  ...           676.0   28     79.99
1   540.0                 0.0      0.0  ...           676.0   28     61.89
2   332.5               142.5      0.0  ...           594.0  270     40.27
3   332.5               142.5      0.0  ...           594.0  365     41.05
4   198.6               132.4      0.0  ...           825.5  360     44.30
5   266.0               114.0      0.0  ...           670.0   90     47.03
6   380.0                95.0      0.0  ...           594.0  365     43.70
7   380.0                95.0      0.0  ...           594.0   28     36.45
8   266.0               114.0      0.0  ...           670.0   28     45.85
9   475.0                 0.0      0.0  ...           594.0   28     39.29

[10 rows x 9 columns]

In [2]:
x = concrete_data.drop(columns = ['Strength'])
y = concrete_data['Strength']

n_cols = x.shape[1]

In [3]:
x.head()

Cement  Blast Furnace Slag  Fly Ash  ...  Coarse Aggregate  Fine Aggregate  Age
0   540.0                 0.0      0.0  ...            1040.0           676.0   28
1   540.0                 0.0      0.0  ...            1055.0           676.0   28
2   332.5               142.5      0.0  ...             932.0           594.0  270
3   332.5               142.5      0.0  ...             932.0           594.0  365
4   198.6               132.4      0.0  ...             978.4           825.5  360

[5 rows x 8 columns]

In [4]:
y.head()

0    79.99
1    61.89
2    40.27
3    41.05
4    44.30
Name: Strength, dtype: float64

In [5]:
n_cols

8

In [6]:
(n_cols,)

(8,)

#A. Build a baseline model (5 marks) 

Use the Keras library to build a neural network with the following:

- One hidden layer of 10 nodes, and a ReLU activation function

- Use the adam optimizer and the mean squared error  as the loss function.



In [7]:
def reg_mod():
  model = Sequential()
  model.add(Dense(10, activation = 'relu', input_shape = (n_cols,)))
  model.add(Dense(1))
  model.compile(optimizer = 'adam', loss = 'mean_squared_error')
  return model

model = reg_mod()

1. Randomly split the data into a training and test sets by holding 30% of the data for testing. You can use the train_test_splithelper function from Scikit-learn.

In [8]:
from sklearn.model_selection import train_test_split

x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.3, random_state = 0)
print(x_train.shape, x_test.shape)


(721, 8) (309, 8)


2. Train the model on the training data using 50 epochs.


In [9]:
model.fit(x_train, y_train, epochs = 50)

Epoch 1/50
23/23 [==============================] - 14s 1ms/step - loss: 165609.5449
Epoch 2/50
23/23 [==============================] - 0s 1ms/step - loss: 91108.8223
Epoch 3/50
23/23 [==============================] - 0s 1ms/step - loss: 45465.8656
Epoch 4/50
23/23 [==============================] - 0s 1ms/step - loss: 21875.1116
Epoch 5/50
23/23 [==============================] - 0s 2ms/step - loss: 9744.9592
Epoch 6/50
23/23 [==============================] - 0s 1ms/step - loss: 4348.8876
Epoch 7/50
23/23 [==============================] - 0s 2ms/step - loss: 2582.0827
Epoch 8/50
23/23 [==============================] - 0s 1ms/step - loss: 2201.4284
Epoch 9/50
23/23 [==============================] - 0s 2ms/step - loss: 2217.4050
Epoch 10/50
23/23 [==============================] - 0s 2ms/step - loss: 2053.9378
Epoch 11/50
23/23 [==============================] - 0s 2ms/step - loss: 1979.9821
Epoch 12/50
23/23 [==============================] - 0s 1ms/step - loss: 1924.6936
Epoch 1

3. Evaluate the model on the test data and compute the mean squared error between the predicted concrete strength and the actual concrete strength. You can use the mean_squared_error function from Scikit-learn.

In [10]:
from sklearn.metrics import mean_squared_error
yhat = model.predict(x_test)
mean_squared_error(y_test, yhat)

563.4124337069984

In [11]:
def evaluate_model(model, x, y, epochs):
  mse = []
  for i in range(50):
    # 1. split the data 
    x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.3, random_state = 0)  
    # 2. train 50 epochs
    model.fit(x_train, y_train, epochs=epochs, verbose=0)
    # 3. mse
    yhat = model.predict(x_test)
    mse.append( mean_squared_error(y_test, yhat) )
    print('.', end='')
  print(' Done!')
  # return the mean and stddev of the mse list
  return np.mean(mse), np.std(mse)


4. Repeat steps 1 - 3, 50 times, i.e., create a list of 50 mean squared errors.

5. Report the mean and the standard deviation of the mean squared errors.

In [12]:
model1 = reg_mod()
mean_mse, std_mse = evaluate_model(model1, x, y, 50)
# Report the mean and stddev of the mean squared errors
print("Mean squared errors for 50 regression models: mean = %.3f, std dev = %.3f" %(mean_mse, std_mse))

.................................................. Done!
Mean squared errors for 50 regression models: mean = 58.958, std dev = 18.390


Submit your Jupyter Notebook with your code and comments.

#B. Normalize the data (5 marks) 

Repeat Part A but use a normalized version of the data. Recall that one way to normalize the data is by subtracting the mean from the individual predictors and dividing by the standard deviation.

In [13]:
x_norm = (x - x.mean()) / x.std()
y_norm = (y - y.mean()) / y.std()

In [14]:
display(x_norm.head())
display(y_norm.head())

Cement  Blast Furnace Slag  ...  Fine Aggregate       Age
0  2.476712           -0.856472  ...       -1.217079 -0.279597
1  2.476712           -0.856472  ...       -1.217079 -0.279597
2  0.491187            0.795140  ...       -2.239829  3.551340
3  0.491187            0.795140  ...       -2.239829  5.055221
4 -0.790075            0.678079  ...        0.647569  4.976069

[5 rows x 8 columns]

0    2.644123
1    1.560663
2    0.266498
3    0.313188
4    0.507732
Name: Strength, dtype: float64

In [15]:
model2 = reg_mod()
mean_mse, std_mse = evaluate_model(model2, x_norm, y_norm, 50)
# Report the mean and stddev of the mean squared errors
print("Mean squared errors for 50 regression models: mean = %.3f, std dev = %.3f" %(mean_mse, std_mse))

.................................................. Done!
Mean squared errors for 50 regression models: mean = 0.145, std dev = 0.015


**How does the mean of the mean squared errors compare to that from Step A?**

**mean mse:**

before - 58.958

after - 0.145

58.958/0.145 = 406


**std mse:**

before - 18.390

after - 0.015

18.390/0.015 = 1226

The improvement of mse after normalization is enormous, it proves that with proper preprocessing Keras models can gain competitive accuracies.


#C. Increate the number of epochs (5 marks)

Repeat Part B but use 100 epochs this time for training.

In [16]:
model3 = reg_mod()
mean_mse, std_mse = evaluate_model(model3, x_norm, y_norm, 100)
# Report the mean and stddev of the mean squared errors
print("Mean squared errors for 50 regression models: mean = %.3f, std dev = %.3f" %(mean_mse, std_mse))

.................................................. Done!
Mean squared errors for 50 regression models: mean = 0.122, std dev = 0.003


**How does the mean of the mean squared errors compare to that from Step B?**

Both mean and std values has improved, the former one by 0.023 and the latter one by 5 times. So it means that the predicted values have become closer to the orifinal target values and the mean error has decreased as well.

#D. Increase the number of hidden layers (5 marks)

Repeat part B but use a neural network with the following instead:

- Three hidden layers, each of 10 nodes and ReLU activation function.


In [17]:
model4 = Sequential()
model4.add(Dense(10, activation='relu', input_shape=(n_cols,)))
model4.add(Dense(10, activation='relu'))
model4.add(Dense(10, activation='relu'))
model4.add(Dense(1))
model4.compile(optimizer='adam', loss='mean_squared_error')

model3 = reg_mod()
mean_mse, std_mse = evaluate_model(model4, x_norm, y_norm, 100)
# Report the mean and stddev of the mean squared errors
print("Mean squared errors for 50 regression models: mean = %.3f, std dev = %.3f" %(mean_mse, std_mse))

.................................................. Done!
Mean squared errors for 50 regression models: mean = 0.130, std dev = 0.005


**How does the mean of the mean squared errors compare to that from Step B?**

The results became even better, the mean mse has improved for 0.015 and, and the std mse has improved 3 times. It demonstrates that adding 2 more hidden layers has improved the model, and it is logical considering the number of inputs in this case.